### PR TITLE
Fix RC Codable to handle arrays and dictionaries from plist defaults

### LINKED ISSUE #9980: RemoteConfig invalid array decoding
### [REQUIRED] Step 1: Describe your environment

  * Xcode version: 13.4.1
  * Firebase SDK version: 9.2.0
  * Installation method: `Swift Package Manage`
  * Firebase Component: `Remote Config`
  * Target platform(s): `iOS`

#### Steps to reproduce:
Example XML:
RemoteConfigDefaults.plist
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
	<dict>
		<key>state</key>
		<string>["yes", "no", "maybe"]</string>
	</dict>
</plist>
```

Fetching the data is fine however decoding runs into the following error:

> DecodingError
  ▿ typeMismatch : 2 elements
    - .0 : Swift.Array<Any>
    ▿ .1 : Context
      - codingPath : 0 elements
      - debugDescription : "Expected to decode Array<Any> but found FirebaseRemoteConfigValueDecoderHelper instead."
      - underlyingError : nil

#### Relevant Code:

```
remoteConfig = RemoteConfig.remoteConfig()
let settings = RemoteConfigSettings()
remoteConfig.configSettings = settings
remoteConfig.setDefaults(fromPlist: "RemoteConfigDefaults")
do {
	config = try remoteConfig.decoded(asType: [String].self)
} catch {
	print(error)
}
```

### PR DESCRIPTION
**Pull request by paulb777**
Fix #9980

The Remote Config Value Codable implementation did not expect arrays or dictionaries other than from json. However, arrays and dictionaries can be specified in defaults set from plists.

This PR adds tests and fixes the implementation.

### CODE REVIEW

File: FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift [hunk 1]
```diff
@@ -47,4 +52,18 @@ struct FirebaseRemoteConfigValueDecoderHelper: FirebaseRemoteConfigValueDecoding
     }
```
Comment (ncooke3) [inline FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift:52]: Should we consider throwing an error here to make behavior more consistent with the API you have added? 

If so, we could create an issue to track this for Firebase 10

File: FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift [hunk 2]
```diff
@@ -47,4 +52,18 @@ struct FirebaseRemoteConfigValueDecoderHelper: FirebaseRemoteConfigValueDecoding
...
   }
+
+  func arrayValue() throws -> [AnyHashable] {
+    guard let value = value.jsonValue as? [AnyHashable] else {
+      throw RemoteConfigValueDecoderError.arrayDecodeFailure
+    }
+    return value
+  }
+
+  func dictionaryValue() throws -> [String: AnyHashable] {
```
Comment (ncooke3) [inline FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift:63]: Are just constrained to `String` keys? Could the plist dictionary have integer keys that are decoded as such?

File: FirebaseRemoteConfigSwift/Tests/Defaults-testInfo.plist [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	<key>arrayIntValue</key>
+	<array>
+		<integer>1</integer>
+		<integer>2</integer>
+		<integer>0</integer>
+		<integer>3</integer>
+	</array>
+	<key>dictValue</key>
+	<dict>
+		<key>foo</key>
```
Comment (ncooke3) [inline FirebaseRemoteConfigSwift/Tests/Defaults-testInfo.plist:34]: My above comment refers to if this was a `1` instead of `"foo"`...

**Review by ncooke3 — APPROVED** [review]
I have a few questions, but this LGTM

File: FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift [hunk 4]
```diff
@@ -18,6 +18,11 @@ import Foundation
 import FirebaseRemoteConfig
 import FirebaseSharedSwift
 
+public enum RemoteConfigValueDecoderError: Error {
+  case arrayDecodeFailure
```
Comment (chliangGoogle) [inline FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift:22]: Just double check for the error should the first letter be capitalized?

Do we need API review for this?

**Review by chliangGoogle — APPROVED** [review]

File: FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift (hunk 1 above)
Comment (paulb777) [reply FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift:52]: Maybe? Returning nil is the historical behavior for failing to extract json. The array and dictionary methods are new and don't have historical precedent.

We might want to wait to we do a more fundamental refactor of RemoteConfigValue.

File: FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift (hunk 2 above)
Comment (paulb777) [reply FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift:63]: I believe the keys must be `String`. See [FirebaseDataEncoder.swift](https://github.com/firebase/firebase-ios-sdk/pull/10046/files#diff-db170d9b3f2582a4ba0a083c21fddd7da24eca78abaa33a0f330a3df8184eb19) line 2479 below.

File: FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift (hunk 4 above)
Comment (paulb777) [reply FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift:22]: Good catch! Removed from the public API surface

File: FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift (hunk 1 above)
Comment (karenyz) [reply FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift:52]: ^ would be interested in discussing this a bit more! feels like `arrayValue` and `dictionaryValue` would be preferred over `jsonValue`

**Review by karenyz — APPROVED** [review]

File: FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift (hunk 4 above)
Comment (paulb777) [reply FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift]: And now completely removed in favor of using the existing DecodingError

File: FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift (hunk 1 above)
Comment (paulb777) [reply FirebaseRemoteConfigSwift/Sources/FirebaseRemoteConfigValueDecoderHelper.swift]: Great point! There's no reason for `jsonValue` any longer and I replaced its usage with `dictionaryValue`.

**Comment by paulb777** [comment]
A future task could be updating `DecodingError._typeMismatch` with more customized error messages for RCValue decoding failures for json, dictionaries and arrays.

**Pull request merged**